### PR TITLE
Deprecate old QNode

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
Currently, the refactored `QNode` is in the `pennylane/beta` folder and users could potentially use it through referencing there. However, maintaining two types of `QNode` objects is not favorable in the long run. Therefore, deprecation of the old-style `QNode` is needed.

**Description of the Change:**
Deprecates the old `QNode` such that only the new version is supported. Moves all related files from the `beta` folder.

**Benefits:**
Only the new `QNode` version needs to be maintained.

**Possible Drawbacks:**
New syntax to be used e.g. for the decorator, several changes need to be noted.

**Related GitHub Issues:**
Series of previous `QNode` PRs, such as https://github.com/XanaduAI/pennylane/pull/420, https://github.com/XanaduAI/pennylane/pull/425

### CODE REVIEW

File: pennylane/qnodes/base.py [hunk 1]
```diff
@@ -477,7 +527,11 @@ def _default_args(self, kwargs):
         # apply defaults
         for name, s in self.func.sig.items():
             default = s.par.default
-            if default != inspect.Parameter.empty:
+            correct_default = all(d != inspect.Parameter.empty for d in default)\
+                              if isinstance(default, np.ndarray)\
+                              else default != inspect.Parameter.empty
+
+            if correct_default:
```
Comment (josh146) [inline pennylane/qnodes/base.py]: @antalszava, were these changes to the logic made to allow a particular set of tests to pass? I'm slightly confused here

File: tests/beta/test_vqe.py [hunk 2]
```diff
@@ -293,21 +294,25 @@ def test_cost_invalid_ansatz(self, ansatz, mock_device):
             cost = qml.beta.vqe.cost([], 4, hamiltonian, mock_device)
 
 
-class TestNumPyInterface:
-    """Tests for the NumPy interface"""
+class TestAutogradInterface:
+    """Tests for the Autograd interface (and the NumPy interface for backward compatibility)"""
```
Comment (josh146) [inline tests/beta/test_vqe.py]: Nice!

File: tests/interfaces/test_tf.py [hunk 3]
```diff
@@ -426,7 +427,7 @@ def circuit(x, y, input_state=np.array([0, 0])):
             qml.RY(y, wires=[0])
             return qml.expval(qml.PauliZ(0))
 
-        circuit = qml.QNode(circuit, qubit_device_2_wires).to_tf()
+        circuit = qml.QNode(circuit, qubit_device_2_wires, interface='tf')
```
Comment (josh146) [inline tests/interfaces/test_tf.py]: This change looks good, how come `to_tf()` doesn't work any more though?

File: tests/test_quantum_gradients.py [hunk 4]
```diff
@@ -458,12 +458,12 @@ def circuit(x, y, z):
         params = np.array([0.1, -1.6, np.pi / 5])
 
         # manual gradients
-        grad_fd1 = qnode.jacobian(params, method='F', order=1)
-        grad_fd2 = qnode.jacobian(params, method='F', order=2)
+        grad_fd1 = qnode.jacobian(params, method='F', options={"order": 1})
+        grad_fd2 = qnode.jacobian(params, method='F', options={"order": 2})
         grad_angle = qnode.jacobian(params, method='A')
 
         # automatic gradient
```
Comment (josh146) [inline tests/test_quantum_gradients.py]:
```suggestion
        # automatic gradient
        # Note: the lambda function is required as evaluate now receives a required `kwargs` argument
        # that cannot be differentiated by autograd.
```

File: tests/test_templates_layers.py [hunk 5]
```diff
@@ -415,15 +387,12 @@ def test_random_layer_weights(self, n_subsystems, tol):
...
+        with qml.utils.OperationRecorder() as rec:
             _random_layer(weights=weights, wires=range(n_subsystems), ratio_imprim=0.3,
                           imprimitive=qml.CNOT, rotations=[RX, RY, RZ], seed=4)
             return qml.expval(qml.PauliZ(0))
 
-        qnode = qml.QNode(circuit, dev)
-        qnode(weights)
-        queue = qnode.queue
-        params = [q.parameters for q in queue]
+        params = [q.parameters for q in rec]
```
Comment (josh146) [inline tests/test_templates_layers.py]: Nice, these tests have been made much simpler! Not only that, but they use fewer areas of the codebase to test the same things.

File: tests/test_templates_subroutines.py [hunk 6]
```diff
@@ -115,24 +106,18 @@ def test_two_mode_rect(self, tol):
...
-            Interferometer(theta, phi, varphi, wires=wires)
-            return [qml.expval(qml.NumberOperator(w)) for w in wires]
 
-        qnode = qml.QNode(circuit, dev)
-        assert np.allclose(qnode(varphi), [0, 0], atol=tol)
-
-        queue = qnode.queue
-        assert len(queue) == 3
+        with qml.utils.OperationRecorder() as rec:
+            Interferometer(theta, phi, varphi, wires=wires)
```
Comment (josh146) [inline tests/test_templates_subroutines.py]: This is also much nicer, in that in removes the unused observable from the test

File: tests/test_templates_subroutines.py [hunk 7]
```diff
@@ -244,24 +214,19 @@ def test_four_mode_triangular(self, tol):
...
+        for idx, op in enumerate(rec.queue[:6]):
             assert isinstance(op, qml.Beamsplitter)
             assert op.parameters == [theta[idx], phi[idx]]
             assert op.wires == expected_bs_wires[idx]
 
-        for idx, op in enumerate(qnode.queue[6:]):
+        for idx, op in enumerate(rec.queue[6:]):
             assert isinstance(op, qml.Rotation)
             assert op.parameters == [varphi[idx]]
             assert op.wires == [idx]
```
Comment (josh146) [inline tests/test_templates_subroutines.py]: Great job fixing these tests!

**Comment by josh146** [comment]
**Update:** I have the docs correctly compiling, and viewable [here](https://pennylane.readthedocs.io/en/deprecate_old_qnode/)

File: pennylane/qnodes/base.py (hunk 1 above)
Comment (antalszava) [reply pennylane/qnodes/base.py]: Exactly. For one of the test cases, a ``np.array`` is passed coming from ``BasisState`` and then ``default != inspect.Parameter.empty`` throws an error.

File: tests/interfaces/test_tf.py (hunk 3 above)
Comment (antalszava) [reply tests/interfaces/test_tf.py]: My bad, both should work equally good. I'll revert his change, thanks!

File: tests/test_templates_layers.py (hunk 5 above)
Comment (antalszava) [reply tests/test_templates_layers.py]: Yes! It's way better to use the ``Recorder``!

**Comment by antalszava** [comment]
That is really nice @josh146 , thanks!

A really nasty caveat that I had to double check and my latest commits correct these cases is exchanging a ``circuit`` just in general with a ``OperationRecorder`` the following way:
```python
def circuit(weights):
    CVNeuralNetLayers(*weights, wires=range(4))
    return qml.expval(qml.X(wires=0))
```
```python
with qml.utils.OperationRecorder() as rec:
    CVNeuralNetLayers(*weights, wires=range(4))
    return qml.expval(qml.X(wires=0))
```
Extra care needs to be taken, as in the second case the test will simply pass due to it reaching the ``return`` statement without *really* checking the test case itself (as most of the times this code snippet would be placed in the beginning of the test).

This is the reason why the ``return`` statements were (and should be) removed when using the ``OperationRecorder``.

Just thought it's worth pointing this out just as a future use of the ``OperationRecorder``.

**Pull request merged**